### DISCUSSION TITLE
[RFC] dma op enhancements

### DISCUSSION THREAD

**Discussion by jgmelber** [Xilinx/mlir-aie#1320] [Ideas]
The `dma` op is a great improvement in syntax over the `dma_star`t op and bd chaining. However it is lacking the `next_bd` feature that is supported in the AIE DMA hardware. It would be great to add this so that more complicated patterns can be expressed. For example bringing in a vector followed by a ping-pong pair to bring in tiles of a larger matrix all on the same DMA channel:

```
@aie.dma(S2MM, 0)
def dma1():
    aie.use_lock(lock_0_1_in_b, AcquireGreaterEqual)
    aie.dma_bd(buffer_0_1_b)
    aie.use_lock(lock_0_1_out_b, Release)
    aie.next_bd(dma1ping)

@aie.another_bd(dma1)
def dma1ping():
    aie.use_lock(lock_0_1_in_a, AcquireGreaterEqual)
    aie.dma_bd(buffer_0_1_a)
    aie.use_lock(lock_0_1_out_a, Release)
    aie.next_bd(dma1pong)

@aie.another_bd(dma2)
def dma1pong():
    aie.use_lock(lock_0_1_in_a, AcquireGreaterEqual)
    aie.dma_bd(buffer_0_1_a)
    aie.use_lock(lock_0_1_out_a, Release)
    aie.next_bd(dma1ping)
```

**Comment by jgmelber** [comment]
@makslevental is this something that could be added?

**Reply by makslevental** [reply]
It's not (currently) possible - this op is implemented using [RegionBranchOpInterface](https://github.com/llvm/llvm-project/blob/f4c0c40f388fff0975ecada4997683cef3cb1fae/mlir/include/mlir/Interfaces/ControlFlowInterfaces.td#L115) which does not (easily) support arbitrary branches. I had some discussions (at the time of implementation) about extending that interface to support but they didn't go anywhere.

**Reply by jgmelber** [reply]
@AndraBisca looked into cleaning up the generated bindings for the `dma_start` op but the syntax either converged to what you have in the new `dma` op or wasn't as clean. It would be great to find a way to support arbitrary branches for DMA BDs.
